internal/controller/apiextensions/composite/api.go

```diff
@@ -428,3 +432,37 @@ func (c *APINamingConfigurator) Configure(ctx context.Context, cp resource.Compo
 	meta.AddLabels(cp, map[string]string{xcrd.LabelKeyNamePrefixForComposed: cp.GetName()})
 	return errors.Wrap(c.client.Update(ctx, cp), errUpdateComposite)
 }
+
+// An APIDeletableChecker checks if composite resource can be deleted.
+type APIDeletableChecker struct {
+	client client.Client
+}
+
+// NewAPIDeletableChecker creates a new APIDeletableChecker.
+func NewAPIDeletableChecker(c client.Client) *APIDeletableChecker {
+	return &APIDeletableChecker{c}
+}
+
+// IsDeleted checks if cr can be deleted (aka have it's finalizer removed).
+//
+// A composite resource can be deleted if all of its composed resources are
+// gone.
+func (c *APIDeletableChecker) IsDeletable(ctx context.Context, cr resource.Composite) (bool, error) {
+	for i, ref := range cr.GetResourceReferences() {
+		nn := types.NamespacedName{
+			Namespace: ref.Namespace,
+			Name:      ref.Name,
+		}
+		obj := unstructured.Unstructured{}
+		// Don't delete the composite if at least one composed resource is still
+		// alive.
+		err := c.client.Get(ctx, nn, &obj)
```

Comment:
Does this work? 🤔 How does the client know what GVK to get? I think you might need to set them on `obj`.